package-res/resources/web/pentaho/visual/ccc/abstract/View.js

```diff
@@ -720,102 +836,92 @@ define([
+     *
+     * Specifies readers (and measuresIndexes and dataCategoriesCount)
+     * because this is the only way to bind specific columns to arbitrary, specific dimensions.
+     * With all the plain table transformations in place,
+     * and apart from the measure discriminator,
+     * which always takes the first position in CCC's logical row,
+     * the CCC readers mapping is 1-1.
      */
     _transformData: function() {
-      var mappedColumnIndexesDiscrete = [],
-          mappedColumnIndexesContinuous = [],
-          C = this._dataView.getNumberOfColumns(),
-          j = -1;
-
-      while(++j < C) {
-        var ai = this._getAttributeInfoByName(this._dataView.getColumnAttribute(j).name);
-        if(ai) {
-          if(ai.attr.isDiscrete)
-            mappedColumnIndexesDiscrete.push(j);
-          else
-            mappedColumnIndexesContinuous.push(j);
-        }
-      }
+      // For DataView source columns
+      var categoriesSourceIndexes = [];
+      var measuresSourceIndexes   = [];
 
-      this._dataView = new DataView(this._dataView);
-      this._dataView.setSourceColumns(mappedColumnIndexesDiscrete.concat(mappedColumnIndexesContinuous));
+      // For CCC readers
+      var categoriesDimNames = [];
+      var measuresDimNames   = [];
+
+      // If there are generic measures, then the cross-tab format is used.
+      if(this._isGenericMeasureMode) {
+        this.options.crosstabMode  = true;
+        this.options.isMultiValued = false;
+
+        categoriesDimNames.push(this.GENERIC_MEASURE_DISCRIM_DIM_NAME);
+
+        this._mappingAttrInfos.forEach(function(mai) {
+          if(mai.attr) {
+            var sourceIndexes = mai.isMeasureGeneric ? measuresSourceIndexes : categoriesSourceIndexes;
+            sourceIndexes.push(mai.attr.ordinal);
+
+            if(!mai.isMeasureDiscrim && !mai.isMeasureGeneric) {
+              categoriesDimNames.push(mai.cccDimName);
+            }
+          }
+        });
 
-      // Set CCC cross-tab mode and categories count
-      // for proper logical row construction.
-      if((this.options.crosstabMode = !!this.measureDiscrimGem)) {
-        this.options.dataCategoriesCount = mappedColumnIndexesDiscrete.length;
+        measuresDimNames.push(this.GENERIC_MEASURE_DIM_NAME);
       } else {
-        // Not relevant, as every position is mapped in readers,
-        // and this only serves to split discrete columns among
-        // series and categories.
-        this.options.dataCategoriesCount = null;
-      }
-    },
+        // The use of the relational translator with the following options
+        // is the most reliable way to "control" the translator code and avoid unexpected shuffling of columns
+        this.options.crosstabMode = false;
+        this.options.isMultiValued = true;
+
+        this._mappingAttrInfos.forEach(function(mai) {
+          if(mai.attr) {
```

Comment:
I find the usage of a short `mai` acceptable in loops like this, because the meaning of `this._mappingAttrInfos` can be understood with minor effort.